### DISCUSSION TITLE
Add upstream certificate to HAR dump output

### DISCUSSION THREAD

**Discussion by cornfeedhobo** [mitmproxy/mitmproxy#6916] [Intercept Everything!]
Is there a way to modify the HAR output format so that it includes the upstream certificate for each entry?

I can see that the feature PR (#6368) deprecated a script, so maybe I can repurpose that approach for a custom HAR format?

**Comment by mhils** [comment]
The dumps should absolutely include that. https://docs.mitmproxy.org/stable/api/mitmproxy/connection.html#Connection.certificate_list on flow.server_conn.

**Reply by cornfeedhobo** [reply]
I was just coming back to comment the same. It appears this is a limitation of the dump format that I used, HAR. I am going to edit the title and question to match that context.

**Comment by mhils** [comment]
You could conceivably monkeypatch SaveHar.flow_entry with a custom addon... If possible, I'd avoid dealing with HAR files though, it's always a mess. 😛

**Reply by cornfeedhobo** [reply]
But it's so ubiquitous! :stuck_out_tongue_closed_eyes:

Can you point the way to using the mitmproxy package as a library to parse native format saves?

**Reply by mhils** [reply]
https://github.com/mitmproxy/mitmproxy/blob/main/examples/addons/io-read-saved-flows.py
